SPDX-License-Identifier: GPL-2.0
aufs6.17 kbuild patch

diff --git a/fs/Kconfig b/fs/Kconfig
index c654a36428970..522605732463c 100644
--- a/fs/Kconfig
+++ b/fs/Kconfig
@@ -334,6 +334,7 @@ source "fs/pstore/Kconfig"
 source "fs/ufs/Kconfig"
 source "fs/erofs/Kconfig"
 source "fs/vboxsf/Kconfig"
+source "fs/aufs/Kconfig"
 
 endif # MISC_FILESYSTEMS
 
diff --git a/fs/Makefile b/fs/Makefile
index 334654f9584b9..292caa46b1ac2 100644
--- a/fs/Makefile
+++ b/fs/Makefile
@@ -130,3 +130,4 @@ obj-$(CONFIG_VBOXSF_FS)		+= vboxsf/
 obj-$(CONFIG_ZONEFS_FS)		+= zonefs/
 obj-$(CONFIG_BPF_LSM)		+= bpf_fs_kfuncs.o
 obj-$(CONFIG_RESCTRL_FS)	+= resctrl/
+obj-$(CONFIG_AUFS_FS)		+= aufs/
