SPDX-License-Identifier: GPL-2.0
aufs5.9 lockdep patch

diff --git a/include/linux/lockdep.h b/include/linux/lockdep.h
index 8e11c3c4d99db..cc77a9692c408 100644
--- a/include/linux/lockdep.h
+++ b/include/linux/lockdep.h
diff --git a/kernel/locking/lockdep.c b/kernel/locking/lockdep.c
index 0d2563b02f18f..97d4c25bf3017 100644
--- a/kernel/locking/lockdep.c
+++ b/kernel/locking/lockdep.c
diff --git a/kernel/locking/lockdep_internals.h b/kernel/locking/lockdep_internals.h
index b0be1560ed17a..6e6d136192def 100644
--- a/kernel/locking/lockdep_internals.h
+++ b/kernel/locking/lockdep_internals.h
@@ -96,15 +96,15 @@ static const unsigned long LOCKF_USED_IN_IRQ_READ =
 #define MAX_STACK_TRACE_ENTRIES	262144UL
 #define STACK_TRACE_HASH_SIZE	8192
 #else
-#define MAX_LOCKDEP_ENTRIES	32768UL
+#define MAX_LOCKDEP_ENTRIES	(32768UL << 5)
 
-#define MAX_LOCKDEP_CHAINS_BITS	16
+#define MAX_LOCKDEP_CHAINS_BITS	(16 + 5)
 
 /*
  * Stack-trace: tightly packed array of stack backtrace
  * addresses. Protected by the hash_lock.
  */
-#define MAX_STACK_TRACE_ENTRIES	524288UL
+#define MAX_STACK_TRACE_ENTRIES	(524288UL << 5)
 #define STACK_TRACE_HASH_SIZE	16384
 #endif
 
